WIP: Dipole moment addition

wandadars opened the pull request:
### Requirements

### Description of the Change

added "dipole_moment" attribute to JSON fluid files for species where the data for the dipole moment is available. The units are in C*m (Where C is Coulomb)


### Benefits

The dipole moment can be used in some thermodynamic calculations.

### Possible Drawbacks

Incorrect value of the dipole_moment

### Verification Process

I don't have access to a code that can output dipole moments for me to compare my changes to.


### Applicable Issues

ibell commented:
Thanks for what you've done so far - looking good.  What's the game plan here?  Collect all of them and then bump the PR?

wandadars commented:
Yeah. I'll continue to add data to the files, and then push more commits for the coding to read the dipole moment data.

wandadars commented:
From the May 3rd response from @ibell on the users group(https://groups.google.com/forum/#!topic/coolprop-users/24yhqpsHcfM), the FluidLibrary.h file should be adjusted to load the dipole moment into the fluid data structure and HEOSMixtureBackend should be modified to return the dipole moments.

I'm trying to understand where those adjustments need to be applied to those two files. I'm adding the dipole moment to the same section as the where the "acentric" factor is located in the JSON fluids files. So for a start I'm thinking about the following additions.

Changed to FluidLibrary.h
(based on line 315, add the following)
EOS.dipole_moment = cpjson::get_double(EOS_json,"dipole_moment");

Changed to HelmholtzEOSMixtureBackend.h
(based on line 196, add the following)
case idipole_moment: return fld.EOS().dipole_moment;

wandadars commented:
Also, is there a standard approach to dealing with species that don't have reported values for the dipole moment?  Do I just not event add a "dipole_moment" attribute?

ibell commented:
Can we merge this and deal with any potential issues later on?  Next step is to make the data accessible through to Python

henningjp commented:
@wandadars Note that two of the fluid JSON files (Cyclopentane.json and Isopentane.json) already have entries for "Dipole_moment_D" which is the dipole moment expressed in unites of Debye.  These appear to be rogues and are never read from the json file.  Your entry with out the "_D" suffix will not conflict, but you may want to remove the older lines in Debye for consistency (or use them - see item 3 below).

Your suggestion above for adding the line to **FluidLIbrary.h** is a good one.  You will also need to create a ``calc_dipole_moment()`` function in **HelmoltzEOSMxtureBackend.h** to overwrite the dummy virtual one in **AbstractState.h**.   You can look at **REFPROPMixtureBackend.cpp** to see how this is done for dipole moment there. 
 
However, ``cpjson::get_double(EOS_json,"dipole_moment")`` will return an error if a fluid file does not have a "dipole_moment" entry.  I might suggest: 

1.  setting a value of, -1 in any fluid files that do not have published value, 
2.  In your "calc_dipole_moment" function, return the value if > 0.  If not, thrown a "Not implemented for this fluid" error.
3.  maybe storing your values in Debye (it's a more manageable number) and multiply it by the conversion factor to C*m before returning it in your ``calc_dipole_moment" function.
4. Perhaps put a comment in the file as to the source of the dipole moment values.

wandadars commented:
Thanks for the feedback @henningjp . I've gone down the road of using C*m units, and I'm committed to this mistake haha. I've been adding the source that was used to inform the commit to the individual commit message for each file(to prevent cluttering the json file). I see now that if we squash those, all that info goes bye-bye. Hopefully I can finish up the process of adding all the attributes this week and then start on adding the coding to process the new entry.

wandadars commented:
What's up with the SulferDioxide JSON file? It looks almost like it contains the same information twice(two entries for "acentric_factor", etc.)

Reference commit: 5da7be164968e1ea683609af1311b71178e153bb

wandadars commented:
@ibell I wrote a script that can express all the dipole moments in Debye and adjust the "dipole_moment_units" json tag from "C*m" to "Debye". I'm just wondering now if that would be a desirable commit on the PR. The Debye isn't an SI unit, but it appears to be the unit most commonly used to express dipole moment measurements. I'm leaning towards doing it because it would make secondary verification easier(others could just immediately see what is written in the json and compare to a different source and see if there are any discrepancies without having to perform any conversions between Debye and C*m. What do you think?

wandadars commented:
> @wandadars Note that two of the fluid JSON files (Cyclopentane.json and Isopentane.json) already have entries for "Dipole_moment_D" which is the dipole moment expressed in unites of Debye. These appear to be rogues and are never read from the json file. Your entry with out the "_D" suffix will not conflict, but you may want to remove the older lines in Debye for consistency (or use them - see item 3 below).
> 
> Your suggestion above for adding the line to **FluidLIbrary.h** is a good one. You will also need to create a `calc_dipole_moment()` function in **HelmoltzEOSMxtureBackend.h** to overwrite the dummy virtual one in **AbstractState.h**. You can look at **REFPROPMixtureBackend.cpp** to see how this is done for dipole moment there.
> 
> However, `cpjson::get_double(EOS_json,"dipole_moment")` will return an error if a fluid file does not have a "dipole_moment" entry. I might suggest:
> 
> 1. setting a value of, -1 in any fluid files that do not have published value,
> 2. In your "calc_dipole_moment" function, return the value if > 0.  If not, thrown a "Not implemented for this fluid" error.
> 3. maybe storing your values in Debye (it's a more manageable number) and multiply it by the conversion factor to C*m before returning it in your ``calc_dipole_moment" function.
> 4. Perhaps put a comment in the file as to the source of the dipole moment values.

Ok @henningjp . I've added -1 to files for fluids where no clear dipole_moment could be found. I added a check for negative dipole_moment values to the "calc_dipole_moment" method that I added to HelmholtzEOSMixtureBackend.cpp. I removed those older "dipole_moment_D" attributes from the JSON files, and in the two places where those attributes were referenced, I adjusted them to ask for just "dipole_moment". I'm planning to convert all dipole moments in the file to Debye like you suggested.

henningjp commented:
@wandadars I think the SO2 file has two equations of state in the file, one from ``Gao-JCED-2016`` and one from ``Lemmon-JCED-2006``.  That's why the constants look like they are listed twice.

henningjp commented:
The equivalent function in the REFPROP interface returns:
``` cpp
            return static_cast<CoolPropDbl>(dip*3.33564e-30);
```
This then converts the Debye units to "C*m" on the fly.  I'll let @ibell weigh in here, but we probably want your function to do this as well because

1. "C*m" is the SI unit
2. It's consistent with the REFPROP backend, whcih converts to "C*m"
4. The documentation already states that the unit returned is "C*m"

Storing it in Debye in the JSON was my point, which you've done, since it is verifiable and a more manageable number in the JSON file (less risk of round-off issues and typos and manual conversion errors).

Your return value would then be:

```cpp
            return components[0].EOS().dipole_moment*3.33564e-30;
```
Since there is the question of units, you might want to check "dipole_moment_units", which you've stored in every case, and do the conversion if it equals "Debye" or "debye" and return it as-is if the unit is "C*m" (and maybe throw and error if it is something unrecognizable).  This would allow future values to be entered in either unit in new fluid JSON files.

The cast to a CoolPropDbl in the REFPROP backend, I think, is superfluous.  All (most?) of the other HEOS calc_ functions return doubles and the value gets type-cast automatically.

ibell commented:
I have no objection at all to the values being stored in the JSON files in
Debye, so long as the values that come out of CoolProp are in C*m.  The
annoying fact is that Debye is a convenient unit (certainly much more so
than C*m), but it's not a base SI unit.  And that's the rule of CoolProp --
all units are base SI, always.  I'm just thankful that someone has done the
work to get these dipole moments into CoolProp.

On Sat, Sep 29, 2018 at 3:07 PM Jeff Henning <notifications@github.com>
wrote:

> The equivalent function in the REFPROP interface returns:
>
>             return static_cast<CoolPropDbl>(dip*3.33564e-30);
>
> This then converts the Debye units to "C*m" on the fly. I'll let @ibell
> <https://github.com/ibell> weigh in here, but we probably want your
> function to do this as well because
>
>    1. "C*m" is the SI unit
>    2. It's consistent with the REFPROP backend, whcih converts to "C*m"
>    3. The documentation already states that the unit returned is "C*m"
>
> Storing it in Debye in the JSON was my point, which you've done, since it
> is verifiable and a more manageable number in the JSON file (less risk of
> round-off issues and typos and manual conversion errors).
>
> Your return value would then be:
>
>             return components[0].EOS().dipole_moment*3.33564e-30;
>
> Since there is the question of units, you might want to check
> "dipole_moment_units", which you've stored in every case, and do the
> conversion if it equals "Debye" or "debye" and return it as-is if the unit
> is "C*m" (and maybe throw and error if it is something unrecognizable).
> This would allow future values to be entered in either unit in new fluid
> JSON files.
>
> The cast to a CoolPropDbl in the REFPROP backend, I think, is superfluous.
> All (most?) of the other HEOS calc_ functions return doubles and the value
> gets type-cast automatically.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CoolProp/CoolProp/pull/1701#issuecomment-425644011>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABxhayUsnkl_r1sfUwI0wzXK4HjfXyI8ks5uf3CEgaJpZM4UVsEc>
> .
>

wandadars commented:
Ok guys. Added the checks that were requested. @henningjp

wandadars commented:
I added a method `calc_dipole_moment` to the source file at _src/Backends/Helmholtz/HelmholtzEOSMixtureBackend.cpp_ .

I added a switch case to the `get_fluid_constant method` in _src/Backends/Helmholtz/HelmholtzEOSMixtureBackend.h_ , that looks like:
`case idipole_moment: return fld.EOS().dipole_moment;`

I also added the following line the the file _src/Backends/Helmholtz/HelmholtzEOSMixtureBackend.h_
`CoolPropDbl calc_dipole_moment(void);`

Travis is coming back with:

> In file included from /Users/travis/build/CoolProp/CoolProp/src/Backends/Cubics/CubicBackend.h:23:
> /Users/travis/build/CoolProp/CoolProp/src/Backends/Helmholtz/HelmholtzEOSMixtureBackend.h:197:51: error: no member named 'dipole_moment' in 'CoolProp::EquationOfState'
>             case idipole_moment: return fld.EOS().dipole_moment;

I thought that this attribute was what came from the coding in the FluidLibrary.h that parsed the dipole moment from the JSON file.

henningjp commented:
@wandadars This looks really good.  Lots of hard work to get all those values in.  I'll try and test in Python in the next few days.  However, if you have any interface (Python, Excel, etc.) that takes calls to PropsSI and/or Props1SI, you should be able to make calls like:
```
Props1SI('DIPOLE_MOMENT', 'SulfurDioxide')
```
To see if it returns the expected values in the right units.

henningjp commented:
@wandadars Fixed the blatant error and it compiles; however something is throwing off indexing of the fluids map in this branch - it can't even find "Water" in the fluids library.  This may take some time to debug and I have other commitments through next weekend, but I'll get to it.

henningjp commented:
@wandadars I'm making progress and will push some changes.  However,

Check your value for **Isopentane** (a.k.a. methylbutane, 2-methylbutane).  Your dipole moment value is 1.0 Debye, but the dipole_moment_D value that you removed from the Chung viscosity model is 0.1 Debye.  I have found multiple other academic references on-line that cite values of 0.12 to 0.13 for Isopentane.  I don't have your cited reference book ["The Properties of Gases and Liquids" 5th edition, Poling, Prausnitz, and O'Connell], but it bears rechecking.  1.0 seems high relative to the other sources.

henningjp commented:
@wandadars - changes PR'd to your base branch.  I think when you accept them, they will show up here.

wandadars commented:
Thank you @henningjp for the commits to fix the code portion of the PR. I have updated isopentane of 0.11, and I noticed that isohexane had a value of 1.0 set, which upon searching some databases looks like it should have been set to -1. So that dipole moment was also updated.

henningjp commented:
@wandadars  - I updated your PR text at the top to include the verification checks.  I only spot checked a few fluids, but the values match the JSON files, zeroes are returned where they should be and fluids with `-1` are returning errors as designed.

@ibell - If you're ok with this, I'll squash and merge.

ibell commented:
Before you squash and merge, can I please run a sanity check on the values
to make sure they agree (at least relatively) with REFPROP?  Or if someone
else could do this test, even better

henningjp commented:
Great idea.  I’ll automate a check against REFPROP for all fluids.

henningjp commented:
@ibell Most values are pretty close.  My computer is going down until Monday, so I'll print a full report later.  @wandadars , just a few look suspect, when compared to the values in REFPROP.  Please recheck references for:

- EthylBenzene (about 30% low)
- n-Butane (400% high)
- NitrousOxide (2.0, looks suspect)
- CycloHexane (0, but I don't think it is)

The rest that you put in look pretty good.  4 out of 122 - Not bad!  CoolProp has a few fluids that REFPROP does not, so there is no comparison value.

henningjp commented:
Ok.  I'm convinced on cyclohexane.  REFPROP lists a very slight dipole moment for cyclohexane, but I don't believe it.  The very weak C-H polarity and the molecule symmetry means it's zero.

henningjp commented:
*ethylbenzene*: I see references for 0.359 D, certainly closer to your value, but I also see references for 0.58 D and 0.59 D (NIST).

henningjp commented:
_Nitrous Oxide_ (N2O) - [NIST](https://cccbdb.nist.gov/diplistx.asp) lists the dipole moment at 0.167 D.  Other sources list similar.

ibell commented:
Since I can't do this easily currently, can you (@henningjp) dump a CSV listing of the new dipole moments and their values in REFPROP into this issue?

henningjp commented:
@ibell Yep. Already created, but my computer is down while I install new floor in my office.  Hope to be back up and running this weekend. (Posting from iPhone)

henningjp commented:
@ibell OK, back up and running.  Value table (csv with a txt extension) comparing to REFPROP results attached.
[DM_Compare.txt](https://github.com/CoolProp/CoolProp/files/2498220/DM_Compare.txt)

ibell commented:
The alkane values in REFPROP seem correct after discussion with @EricLemmon
and searching some references online.  Where did the values in CoolProp
come from?

henningjp commented:
@ibell Should we create an additional dipole_moment_source key that holds the reference source string right in the JSON file?  I think if it is not accessed by the code it would be a valuable comment in each file as to where the values come from, especially since there is some discrepancy between sources.  Would that extra key be innocuous?

henningjp commented:
@wandadars , you or me?  I don't want to duplicate effort if you're already making these keys for the source string in each JSON.  Either way is fine with me.

wandadars commented:
@henningjp  I haven't started any work on adding the additional keys. I'm busy with some other things this week, so I won't be able to get started anytime soon if you want to take a crack at adding those keys.

henningjp commented:
@wandadars Sorry for the delay.  It's a long sad story of corrupted local git repositories, lost files, and having to repeat the work all over again.  I made a new PR to your branch that contains the following:

1. Added the source references from your original commits to the "dipole_moment_ref" field in each of the fluid JSON files.
2. Fixed some minor precision issues with some of the values.  There were some extra digits that weren't in the source reference.  These were probably rounding residue from the conversion to C*m and back to Debye units.  Sorry for that unnecessary iteration.
3. Figured out the issue with the Alkanes.  Back when these were originally committed, I checked the JSON values against eThermo and they indeed had the values posted online.  Checking this week they have changed!  The values posted were actually the acentric factor and they are now labelled as such on the eThermo site.  New values that much more closely match the REFPROP values are now posted for dipole moment for all the normal alkanes.
4. Found a source reference for a few refrigerants that were in REFPROP but missing from CoolProp.

To verify everything I took the following steps:

- [x] Compiled and ran CoolProp with the new JSON files and all fluids load successfully and report dipole moment values.
- [x] Generated a new comparison csv file between the new CoolProp values and the values reported from REFPROP, listing the % difference between the two.  This csv is attached here (Github made me rename it to a .txt file).  The list is sorted by %difference from REFPROP (low to high) and I put notes in the fields that have suspect values or missing data.

[DM_Compare_sorted.txt](https://github.com/CoolProp/CoolProp/files/2631805/DM_Compare_sorted.txt)

henningjp commented:
@wandadars Here are a few remaining precision issues to check against Poling and two with the Geleste references.  They are nit-picky, but they should be double checked.  If the Poling values are accurate, then I'm fine with them as the difference is small and they are newer than the CCDBD and NBS10 references used heavily by REFPROP.  Thanks.

Fluid | CP Value | REFPROP Value | Comment
------|-----------|------------------|------------
IsoButene | 0.5001 | 0.5 | Check Value precision in Poling reference
cis-2-Butene | 0.3001 | 0.3 | Check Value precision in Poling reference
DimethylEther | 1.3 | 1.301 | Check Value   precision in Poling reference
D4 | 1.091 | 1.09 | 1.09 reported by https://www.gelest.com/product/octamethylcyclotetrasiloxane-98/ This should be updated.
Ethanol | 1.7 | 1.6909 | Check Value   precision in Poling reference
R12 | 0.5001 | 0.51 | NSRDS-NBS19 lists   0.51 ; Recheck Poling reference
HydrogenChloride | 1.1 | 1.079 | Check Value precision in Poling reference; NIST CCBDB and NSRDS-NBS10 report 1.08

henningjp commented:
@wandadars , @ibell suggested a target of 1% difference from the REFPROP values; however, this may not be reasonable given some of the very small values.  A survey of the fluids that do not agree with REFPROP are listed in the tables below with comments and a suggested resolution. (NOTE: the rest of the fluids in the library agree 100% with the REFPROP values.)  For a few of the fluids in the table, I land on the side of the submitted values (newer, confirmed references over the REFPROP source) and a few (???) have non-reconciled differences.  References are only used that we can actually lay eyes on and I do not have access to DIPPR, which is used for several of the REFPROP values.  If anyone has issue with the proposed resolutions, or has suggestions on the non-reconciled differences, or can provide a better reference value, please comment.

Fluid Name | CoolProp | REFPROP | Comment | Resolution
-----------|----------|---------|---------|-----------
CarbonylSulfide   | 0.712 | 0.7152 | NIST CCCBDB list (on-line) and NSRDS-NBS10 (1967) list   0.712;     REFPROP: J.S. Muenter, J. Chem. Phys., 48, 4544 (1968) (0.7152); From   Muenter "The value obtained is 0.71521 ± 0.00020 D, which is in   disagreement with a previous measurement of 0.7124 ± 0.0002 D." | Use   the Muenter value
Ethanol | 1.7 | 1.6909 |   Poling (1.7), but rounds to single decimal place;     NSRDS-NBS10 (1.69);     REFPROP references DIPPR | Use NSRDS-NBS10 (1.69)
R12 | 0.5001 | 0.51 |   NSRDS-NBS10 lists 0.51 ;     Poling (0.5); (but all values rounded to single decimal place)     REFPROP 9/10 reference REFPROP 5 (!!!) | Use NSRDS-NBS10 (0.51)
HydrogenChloride | 1.1 | 1.079   | NIST CCBDB and NSRDS-NBS10 report 1.08     REFPROP references DIPPR;     Poling has 1.1, but rounds all values to single decimal. | Use NSRDS-NBS10   (1.08)
n-Propane | 0.083 | 0.084 | J.   Chem. Phys. 33, 1514 (1960); https://doi.org/10.1063/1.1731434;      NBS10 cites this reference (0.83) and averages with Muenter, J. S. and   Laurie, V. W.:     J. Chem. Phys. 45, 855 (1966). (0.85) to get 0.84.  This is probably the best value. | Use   NSRDS-NBS10 value (0.084 D) which is slightly newer.
MM | 0.7801 | 0.801 | 0.78   reported by https://www.gelest.com/product/hexamethyldisiloxane-98/ Maybe   update, but REFPROP value comes from DIPPR; not sure what to do on this one.   | Close enough?
DimethylCarbonate | 0.93 |   0.899 | Dimethyl Carbonate (C3H6O3).    Various cis/trans conformers make this one hard to pin down and find a   source.  Not listed in Poling/Reid.  The REFPROP value comes from DIPPR.     NBS10 states "The significance of these values may involve some   ambiguity because of the possibility of different conformations or spatial   isomers." and does not report a value but references the paper   below:     Kubo, M., Morino, Y. and Mizushima, S.: Sci. Papers Inst. Phys. Chem. Res.   (Tokyo) 32, 129-37 (1937) (0.9 @325K, 1.05 @497K).     Tisato Kajiyama, Polymer Science and Industrial Research in the Fast-changing   Age: 7th SPSJ International Polymer Conference (IPC99) (1.0)     Tundo, P.; Selva, M. The Chemistry of Dimethyl Carbonate. Acc. Chem. Res.,   2002, 35 (9), pp 706–716     DOI: 10.1021/ar010076f (0.91) | Use the Tundo value (0.91 D)
NitrousOxide | 0.167 | 0.1608   | NIST CCBDB and NSRDS-NBS10 report 0.167 ; Poling rounds to 0.2;      REFPROP references J. Chem. Phys. 53, 2513 (1970);   https://doi.org/10.1063/1.1674355 | Use NBS10 (0.167 D) - OK
1-Butene | 0.359 | 0.339 |   NIST CCCBDB list (on-line) lists 0.359. Vector sum of µ(a), µ(b), µ(c) gives   0.359(3).     Poling/Reid (0.3) [rounded];     NSRDS-NBS10 (0.34);  REFPROP   references this source. | Use NSRDS-NBS10 (0.34 D)
HFE143m | 2.32 | 2.48 |   REFPROP: (computed by A. Kazakov, Nov. 2017, DF-MP2/def2-QZVPD, unc. 0.1-0.15   D);     ethermo (2.32);     Gage, C. L. and Kazachid, O. S., "Predictions of Azeotropes Formed   from Fluorinated Ethers, Ethanes, and Propanes" (1992), International   Refrigeration and Air Conditioning Conference. Paper 201. (2.45) | Use   Gage/Kazachid (2.45 D)
HydrogenSulfide | 0.9 | 0.97 |   Poling (0.9);      REFPROP: NIST CCBDB and NSRDS-NBS10 (0.970) | Use NSRDS-NBS10 (0.970 D)
MD2M | 1.22 | 1.12 |      Decamethyltetrasiloxane     Commercial site https://www.gelest.com/product/decamethyltetrasiloxane/   reports 1.22.  No other source   found.  decamethyltetrasiloxane.     REFPROP value from DIPPR | 1.22 is best number we have
D5 | 1.22 | 1.349 |   decamethylcyclopentasiloxane;     Gelest @ https://www.gelest.com/product/decamethylcyclopentasiloxane/   (1.22);      http://www.ethermo.us/Mars905.htm (1.349)     REFPROP: DIPPR (1.349) | use eThermo (1.349)
R11 | 0.5001 | 0.45 | R11   (CFCl3) is listed in NIST CCCBDB and NSRDS-NBS10 as 0.45;     Poling lists 0.5 (newer, but all values rounded to single decimal   place);     REFPROP 9/10 reference REFPROP 5 (!!!) probably from NSRDS-NBS10. | Use   NBS10 value (0.45)
CarbonMonoxide | 0.112 | 0.1 |   NIST & NSRDS-NBS10 list 0.112(0) references Microwave Spectoscopy method   reported in Burrus, C. A.:J. Chem. Phys. 28, 427-9 (1958).  Both Poling and Reid report 0.1. | Leave   alone or reference Poling?
DiethylEther | 1.31 | 1.151   | Poling (1.3);  Reid (1.3);      REFPROP Version 9/10 from DIPPER/DIADEM (1.151)($);     Lange's Hbk (1.15 D)     NIST CCCBDB and NSRDS-NBS10 (1.15);     J. Am. Chem. Soc. 1950,  72, 10,   4832-4832 (1.28)     Dortmund Data Bank (http://www.ddbst.com/en/EED/PCP/DPM_C12.php) reports   1.45 D (liquid) | Use NSRDS-NBS10 (1.15)
n-Hexane | 0.08 | 0.07 | Close   enough given the very small magnitude of these values \| OK
Novec649 | 0.36 | 0.43 |   eThermo site (0.36 D);     REFPROP cites: calculated by A. Kazakov, April 2017, unc. +/- 0.12-.2   (0.43)     Not sure what to do with this one. No other sources. | ???
MDM | 0.8001 | 0.99 |   C8H24O2Si3     Commercial site https://www.gelest.com/product/octamethyltrisiloxane/  (0.80 D).         Gelest also reports 98% Concentration value (1.09 D)     http://www.ethermo.us/ShowDetail93.htm (1.079 D)     Thermophysical Properties of Chemicals and Hydrocarbons, By Carl L. Yaws   (1.079 D)     REFPROP: Sauer, R.O. and Mead, D.J., J. Am. Chem. Soc., 68(9):1794-1797,   1946. (0.99 D) | Use Yaws (1.079 D)
EthylBenzene | 0.3999 | 0.6 |   NIST CCBDB and NSRDS-NBS10 (0.590 D)     REFPROP: DIPPR (0.60 D)     Poling / Reid (0.4 D)     http://www.vias.org/genchem/dipole_moment_table.html (0.36 D)     http://www.ethermo.us/mars1041.htm (0.4)     Journal of Chromatographic Science, Vol. 43, February 2005, Hiromitsu   Kanai, et. al., pp. 57- (0.37) | Majority point to values near 0.4; Use   Poling (0.4 D)
CycloHexane | 0 | 0.3 |   Should be a symmetric non-polar molecule, and there are many references that   list 0.0 D.  However,      REFPROP references Poling (0.3)     Reid (0.3)     J. Chem. Phys. 49, 2368 (1968); https://doi.org/10.1063/1.1670409   (0.331)     Cyclohexane has a stable "chair" form (symmetric) and a boat   conformer (non-symmetric).  At 25°C it   exists as 99.9% chair form.  This would   indicate a dipole moment of zero, but at higher energy levels some % of the   conformer could lead to slightly higher values. | Use Poling (0.3 D)
n-Dodecane | 0.07 | 0 |   Original CP reference (ethermo.us/hydrocarbon54) no longer shows a dipole   moment value;     REFPROP: Dornte, R.W. and C.P. Smyth, J. Am. Chem. Soc. 52, 3546-3552   (1930) (0.0 D);     Poling (0.0 D);     http://www.ethermo.us/ShowDetail61.htm (0.0D) - no reference;     http://www.stenutz.eu/chem/solv6.php?name=dodecane (0.07 D) - no reference;   | Use Poling (0.0 D)
R218 | 0.14 | 0 |   (Octofluoropropane)     Poling (0.0 D);     REFPROP: ab-initio calculations from HF 3-21G (0.0 D);     http://www.ethermo.us/ShowDetail52.htm (0.07 D - venus model);     http://www.ethermo.us/mars1546.htm (0.0 D - mars model); | Use Poling (0.0   D)
Air | -1 | 0 | Mixture   unsupported;  If we use the fact that   air is mostly O2 and N2 and that these are monatomic molecules, then 99.9% of   air has a dipole moment of zero. | Either -1 (undefined) or 0
Isohexane | -1 | 0 | (alias   2-methylpentane)     REFPROP: ab-initio calculations from HF 6-31G* (0.0 D);     Poling (0.1 D);     Lange's (0.0 D);     http://www.stenutz.eu/chem/solv6.php?name=2-methylpentane reports 0.00D |   Use Lange's
MD3M | -1 | 1.223 | REFPROP:   DIPPR (1.223 D);     Comercial - https://www.gelest.com/product/dodecamethylcyclohexasiloxane/   (1.56 D); | ???
MD4M | -1 | 1.308 |   (Tetradecamethylhexasiloxane)     REFPROP: DIPPR (1.308 D);     Lange's Hbk (1.58 D) | ???
MethylLinoleate | -1 | 1.79 |   REFPROP: DIPPR ;      http://www.ethermo.us/ShowDetail100.htm (-1) - unknown;     No other sources found. | OK
R365MFC | -1 | 3.807 |   (Solkane 365)       https://www.solvay.com/sites/g/files/srpend221/files/tridion/documents/SOLKANE-365-Green-Solvent.pdf   (3.8);     REFPROP: DIPPR (3.807) | Use Solvay site (3.8)

**Fluids with no dipole moment values listed in REFPROP**

Fluid Name | CoolProp | REFPROP | Comment | Resolution
-----------|----------|---------|---------|-----------
Dichloroethane   | 1.799 | -1 | (1,2-Dichloroethane)     Fluid not in Refprop 9.1.1 or 10.x;     Poling (1.8 D);     Reid (1.8 D);     NBS10 does not list a value; | Use Poling (1.8 D)
EthyleneOxide | 1.94 | -1 |   Not in Refprop 9.1.1 or 10.x;      NIST CCBDB and NSRDS-NBS10 (1.89 D);     CRC Handbook of Chemistry and Physics, 92nd ed. (1.94 D);     Reid (1.9 D);  | Use CRC - OK
OrthoDeuterium | 0 | -1 | spin   isomers still monatomic; Same as Deuterium; Fluid not in REFPROP | Use zero value
ParaDeuterium | 0 | -1 | spin   isomers still monatomic; Same as Deuterium; Fluid not in REFPROP | Use zero value
R1233zd(E) | 1.44 | -1 |   ethermo http://www.ethermo.us/ShowDetail2820.htm only known source (1.44) |   OK
R1234ze(E) | 1.27 | -1 |        Alt Reference: https://nvlpubs.nist.gov/nistpubs/jres/117/jres.117.014.pdf   (1.27) - calculated | Use NIST Source (1.27)

Once we're agreed on the resolutions above, I'll add to the PR with the above resolutions.

henningjp commented:
@ibell  - What do you want to do with this?   I can wrap it up, but I think it needs a completely new PR as it's quite outdated.  I think I've got reasonable assessments of the values (what to include, what to leave undefined) in the previous comments.  However, I've not heard anything back from you on direction in a very long time.  Is this worth finishing up?

ibell commented:
@henningjp - It would indeed be great to get this into CoolProp.  Sorry I've been not as responsive on this thread as I might like. I agree that a fresh PR might be the way to go given its age.

ibell commented:
I don't think I can dig into the nitty gritty of the values where diescrepancies occur, but where we are all in agreement about what the value should be, let's get those in.

Also, this quantity is much mushier than it seems, so I think my 1% target was likely ambitious.  Maybe more like 5% would do.
